src/sources/datadog/agent/mod.rs

```diff
@@ -159,13 +161,38 @@ impl SourceConfig for DatadogAgentConfig {
     }
 
     fn outputs(&self) -> Vec<Output> {
+        let definition = match self.decoding {
+            // See: `LogMsg` struct.
+            DeserializerConfig::Bytes => schema::Definition::empty()
+                .required_field("message", Kind::bytes(), Some("message"))
+                .required_field("status", Kind::bytes(), Some("severity"))
+                .required_field("timestamp", Kind::integer(), Some("timestamp"))
+                .required_field("hostname", Kind::bytes(), Some("host"))
+                .required_field("service", Kind::bytes(), None)
+                .required_field("ddsource", Kind::bytes(), None)
+                .required_field("ddtags", Kind::bytes(), None)
+                .merge(self.decoding.schema_definition()),
+
+            // JSON deserializer can overwrite existing fields at runtime, so we have to treat
+            // those events as if there is no known type details we can provide, other than the
+            // details provided by the generic JSON schema definition.
+            DeserializerConfig::Json => self.decoding.schema_definition(),
+
+            // Syslog deserializer allows for arbritrary "structured data" that can overwrite
+            // existing fields, similar to the JSON deserializer.
+            //
+            // See also: https://datatracker.ietf.org/doc/html/rfc5424#section-6.3
+            #[cfg(feature = "sources-syslog")]
+            DeserializerConfig::Syslog => self.decoding.schema_definition(),
```

Comment:
It's important to note here that we intend to change the behaviour of `message` deserialization in the future, by nesting the deserialized data into their own field (e.g. `attributes`, or `syslog`). This would allow operators to use a decoder, _and_ still have the exact type information exposed by the relevant source.

This was deemed out-of-scope for now, but we'll revisit this soon.